botocore/client.py

```diff
@@ -325,6 +246,144 @@ def _api_call(self, *args, **kwargs):
         return _api_call
 
 
+class ClientEndpointBridge(object):
+    """Bridges endpoint data and client creation
+
+    This class handles taking out the relevant arguments from the endpoint
+    resolver and determining which values to use, taking into account any
+    client configuration options and scope configuration options.
+
+    This class also handles determining what, if any, region to use if no
+    explicit region setting is provided. For example, Amazon S3 client will
+    utilize "us-east-1" by default if no region can be resolved."""
+
+    SIGNATURE_PRECEDENCE = ['s3', 's3v4', 'v4', 'v2']
```

Comment:
It's both a future proof thing and an easy way to define which signers we support in Botocore. If we find one of the preferred signers in the list of signature versions, then we know that we can send requests to a service.

Checking in a predefined order is a future proof thing. The reason to check for v4 and prefer it over v2 is to correctly handle the case of one of the few remaining v2 services adding support for v4. At that point, it is unlikely the service would stop supporting v2, so we'd need to account for the case of both v2 and v4 as available signature versions.